### PR TITLE
`click.get_pager_file`: add tests (#1572 followup)

### PR DESCRIPTION
**Pull request by AndreasBackx**
This is a follow-up to #1572 which adds tests to validate the behaviour of `click.get_pager_file`. There are still some open-standing questions / todos depending on requirements:

- No Windows tests (yet) because I'm visiting family right now and only have access to a Linux machine.
	- [ ] What pager like a simple `cat` could we use to create simple tests? We might be able to install `cat` on our CI, though I haven't looked into it yet.
- What are some other edge cases we could check for? So far I've only added some simple tests:
	- Simple unit tests to test some internal implementation details. I'm not the happiest with this, but it seems like the termui testcode favours testing some internals and it was the easiest of the bunch to add.
	- Simple `cat`-based integration tests. We basically want to make sure that it works how we expect with a real pager, but using `cat` seems like the easiest way to validate this as I don't know how we'd test interactive CLIs. :thinking: 
	- [ ] Decide whether the current list of tests are complete enough or if we're missing anything, particularly edge cases.

### CODE REVIEW

**Review by kdeldycke — CHANGES_REQUESTED** [review]
Can you add reference to that PR in the changelog along side the one from #1572?

i.e.:
```rst
-   Add ``click.get_pager_file`` for file-like access to an output
    pager. :pr:`1572` :pr:`3405`
```

Also update the `.. versionadded:: 8.2` that was left-over from #1572.

**Review by kdeldycke — DISMISSED** [review]
There are workaround and late changes in #1572 about windows so yes, it's better to try to test on that platform too. If you don't have access to a Windows machine, just remove your skip decorator and use the feedback from GitHub workflow runs. Slow, but doable.

File: tests/test_termui.py [hunk 1]
```diff
@@ -620,6 +626,160 @@ def test_pager_shlex_split(pager_env, expected_parts):
...
+        pager.write("suffix\n")
+
+    output = _run_get_pager_file_with_real_pager(monkeypatch, tmp_path, writer)
+
+    assert output == "prefix\nmiddle\nsuffix\n"
+
+
+@pytest.mark.skipif(WIN, reason="Exercises the pipe pager path.")
+@pytest.mark.parametrize(
+    ("text", "color", "expected"),
```
Comment (kdeldycke) [inline tests/test_termui.py]: I like this matrix of ANSI test cases! 👍

File: tests/test_termui.py [hunk 2]
```diff
@@ -620,6 +626,160 @@ def test_pager_shlex_split(pager_env, expected_parts):
...
+    """An already-wrapped pager stream should be yielded unchanged."""
+    buffer = io.BytesIO()
+    wrapped = click._termui_impl.MaybeStripAnsi(buffer, color=True, encoding="utf-8")
+
+    @contextlib.contextmanager
+    def pager_contextmanager(color=None):
+        yield wrapped, "utf-8", color
+
+    monkeypatch.setattr(
+        click._termui_impl, "_pager_contextmanager", pager_contextmanager
```
Comment (kdeldycke) [inline tests/test_termui.py]: Instead of monkeypatching `_pager_contextmanager` everytwhere, you can just feed `get_pager_file` with `PAGER=`. That way you can check with `PAGER` being set to `cat`, `PAGER` left unset, `PAGER` set to a missing binary etc to test even more edge-cases.

File: tests/test_termui.py [hunk 3]
```diff
@@ -620,6 +626,160 @@ def test_pager_shlex_split(pager_env, expected_parts):
...
+    monkeypatch.setattr(click._termui_impl, "isatty", lambda _: True)
+    monkeypatch.setitem(click._termui_impl.os.environ, "PAGER", pager_path)
+
+    with pager_out.open("w") as f:
+        force_subprocess_stdout = patch.object(
+            subprocess,
+            "Popen",
+            partial(subprocess.Popen, stdout=f),
+        )
+        with force_subprocess_stdout:
```
Comment (kdeldycke) [inline tests/test_termui.py]: Just a style nitpicking: why do you need the temporary `force_subprocess_stdout` variblae? Just fold everything in the with statement.

File: tests/test_termui.py (hunk 2 above)
Comment (kdeldycke) [reply tests/test_termui.py]: Yes, I think that's the way to go because this is the top level API that is being modified by #1572. So if there are regressions, they will surface from here. Maybe you can even repack your tests to only target that higher-level API and demonstrate more use-cases, and skip testing implementation details.

And you can do the same with the other `echo_via_pager` function that was impacted by #1572.

File: tests/test_termui.py [hunk 4]
```diff
@@ -620,6 +626,160 @@ def test_pager_shlex_split(pager_env, expected_parts):
...
+            with click.get_pager_file(color=color) as pager:
+                writer(pager)
+
+    return pager_out.read_text()
+
+
+@pytest.mark.skipif(WIN, reason="Exercises the pipe pager path.")
```
Comment (kdeldycke) [inline tests/test_termui.py]: I just learned the existence of the `more` utility on Windows. So you can probably test with that as a `PAGER`.

File: tests/test_termui.py (hunk 3 above)
Comment (AndreasBackx) [reply tests/test_termui.py]: Explains a bit what it does and makes the with statement a bit less unsightly, just a preference and I don't think too much about it in test code.

File: tests/test_termui.py (hunk 4 above)
Comment (AndreasBackx) [reply tests/test_termui.py]: I've given more a try. Let's see if it passes.

File: tests/test_termui.py (hunk 2 above)
Comment (AndreasBackx) [reply tests/test_termui.py]: I've redesigned it all a bit because I made the mistake of diving into the differences between BinaryIO, TextIO, io.StringIO, and other io shenanigans. My brain is a bit fried and I'm just going to make dinner as I'm starving.

Let me know what you think. I don't think it's perfect yet, but honestly am just gonna put this down for today and give it another look another day.

**Review by AndreasBackx — COMMENTED** [review]
Will still update the change entries, but likely not today.

**Comment by kdeldycke** [comment]
Ah. You're right. Windows runners are failing hard on `more`. That's probably because `more` is interactive and waiting user input. I guess it is safe to skip on Windows given the circumstances.

I also spotted some opportunities to extend the tests with `@parametrize` decorator. Let me add a commit on top of your changes to show you what I mean.

**Comment by kdeldycke** [comment]
@AndreasBackx I added a couple of tests around the `PAGER` env var, updated some references and re-introduced the Windows skipping behavior you originally implemented.

What do you think of these changes?

**Comment by Rowlando13** [comment]
I don't want to release 8.4.0 until these go in. @kdeldycke Feel free to keep adding small fixes and I will merge until this is ready. However, when this is ready for merge, I will accept it and just cut the release, but I expect 8.4.1 to come out shortly so I don't expect the prs to sit for long.

**Comment by kdeldycke** [comment]
> I don't want to release 8.4.0 until these go in. @kdeldycke Feel free to keep adding small fixes and I will merge until this is ready. However, when this is ready for merge, I will accept it and just cut the release, but I expect 8.4.1 to come out shortly so I don't expect the prs to sit for long.

Yes, you are right to push for stale PRs and prioritize coverage now. We need a bit of effort to arrive at a final 8.4.0.

**Comment by kdeldycke** [comment]
> I don't want to release 8.4.0 until these go in. @kdeldycke Feel free to keep adding small fixes and I will merge until this is ready. However, when this is ready for merge, I will accept it and just cut the release, but I expect 8.4.1 to come out shortly so I don't expect the prs to sit for long.

Just rebased on top of `main`. This is ready to be merged upstream.
